app/app.js

```diff
@@ -0,0 +1,233 @@
+
+import {AsyncStorage} from 'react-native';
+import DeviceInfo from 'react-native-device-info';
+
+import {loadMe} from 'mattermost-redux/actions/users';
+import EventEmitter from 'mattermost-redux/utils/event_emitter';
+
+import {ViewTypes} from 'app/constants';
+import mattermostManaged from 'app/mattermost_managed';
+import tracker from 'app/utils/time_tracker';
+
+import {store, handleManagedConfig} from 'app/mattermost.android.js';
+
+const lazyLoadLocalization = () => {
+    const IntlProvider = require('react-intl').IntlProvider;
+    const getTranslations = require('app/i18n').getTranslations;
+
+    return {
+        IntlProvider,
+        getTranslations
+    };
+};
+
+const DEVICE_SECURE_CACHE_KEY = 'DEVICE_SECURE_CACHE_KEY';
+const DEVICE_TOKEN_CACHE_KEY = 'DEVICE_TOKEN_CACHE_KEY';
+const USER_ID_CACHE_KEY = 'USER_ID_CACHE_KEY';
+const TOKEN_CACHE_KEY = 'TOKEN_CACHE_KEY';
+const URL_CACHE_KEY = 'URL_CACHE_KEY';
+
+export default class App {
+    constructor() {
+        // Usage: app.js
+        this.shouldRelaunchWhenActive = false;
+        this.inBackgroundSince = null;
+        this.nativeAppLaunched = false;
+
+        // Usage: screen/entry.js
+        this.startAppFromPushNotification = false;
+        this.isNotificationsConfigured = false;
+        this.allowOtherServers = true;
+        this.appStarted = false;
+        this.emmEnabled = false;
+        this.intl = null;
+
+        // Usage utils/push_notifications.js
+        this.replyNotificationData = null;
+        this.deviceToken = null;
+
+        // Usage credentials
+        this.currentUserId = null;
+        this.token = null;
+        this.url = null;
+
+        this.getAppCredentials();
+    }
+
+    getIntl = () => {
+        if (this.intl) {
+            return this.intl;
+        }
+        const {
+            IntlProvider,
+            getTranslations
+        } = lazyLoadLocalization();
+
+        const state = store.getState();
+        let locale = DeviceInfo.getDeviceLocale().split('-')[0];
+        if (state.views.i18n.locale) {
+            locale = state.views.i18n.locale;
+        }
+
+        const intlProvider = new IntlProvider({locale, messages: getTranslations(locale)}, {});
+        const {intl} = intlProvider.getChildContext();
+        this.setIntl(intl);
+        return intl;
+    };
+
+    getManagedConfig = async () => {
+        try {
+            const shouldStart = await AsyncStorage.getItem(DEVICE_SECURE_CACHE_KEY);
+            if (shouldStart !== null) {
+                return shouldStart === 'true';
+            }
+        } catch (error) {
+            return false;
+        }
+        return false;
+    };
+
+    getAppCredentials = async () => {
+        try {
+            const [
+                deviceToken,
+                currentUserId,
+                token,
+                url
+            ] = await Promise.all([
+                await AsyncStorage.getItem(DEVICE_TOKEN_CACHE_KEY),
```

Comment:
Why are we using this directly from the AsyncStorage and not from what we have in redux?

Reply:
will it cause you to logout when you upgrade app to this version?
also wil it be properly cleaned up on logout?